test/spec/functions-openrosa.spec.js

```diff
@@ -1068,4 +1065,36 @@ describe( 'Custom "OpenRosa" functions', () => {
 
     } );
 
+    describe('digest', () => {
+      it( 'digest', () => {
+        [
+          ["digest('abc', 'SHA-1', 'hex')", 'a9993e364706816aba3e25717850c26c9cd0d89d'],
+          ["digest('abc', 'SHA-256', 'hex')", 'ba7816bf8f01cfea414140de5dae2223b00361a396177a9cb410ff61f20015ad'],
+          ["digest('abc', 'SHA-256')", 'ungWv48Bz+pBQUDeXa4iI7ADYaOWF3qctBD/YfIAFa0='],
+          ["digest('abc', 'SHA-256', 'base64')", 'ungWv48Bz+pBQUDeXa4iI7ADYaOWF3qctBD/YfIAFa0=']
+        ].forEach( async ([expr, expected]) => {
+          var result = g.doc.evaluate(expr, g.doc, null, g.win.XPathResult.STRING_TYPE, null);
+
+          // The web crypto api only supports async functions and does not support md5
+          // https://www.w3.org/TR/WebCryptoAPI/#SubtleCrypto-method-digest
+          // https://www.w3.org/2012/webcrypto/track/issues/24
+          // Is this a weird promise?
+          expect(await (result.stringValue)).to.equal(expected);
+        });
+
+        const invalidAlgoTest = () => {
+          g.doc.evaluate("digest('abc', 'XYZ', 'hex')", g.doc, null, g.win.XPathResult.STRING_TYPE, null);
+        }
+
+        expect(invalidAlgoTest).to.throw();
+
+        const invalidEncodingTest = () => {
+          g.doc.evaluate("digest('abc', 'SHA-1', 'x')", g.doc, null, g.win.XPathResult.STRING_TYPE, null);
+        }
+
+        expect(invalidEncodingTest).to.throw();
+      } );
+    } );
+
+
```

Comment:
Unfortunately, I don't think we can do asynchronous functions in XPath. I am also not very enthusiastic about adding a large library to do this synchronously, so I'm wondering if we should just forget about supporting `digest()`. It's quite an obscure function.

We do include a library in enketo-express called [`node-forge`](https://github.com/kobotoolbox/enketo-express/blob/master/package.json#L68), which would not increase the size (if the versions are in sync?) in Enketo Express. Not sure if that would support synchronous digest generation.